### PR TITLE
#7049: Test yolov4 Downsample2 and Downsample3 ttnn sub-modules

### PR DESCRIPTION
**Pull request by saichandax**
This is a draft PR for the downsample2 and downsample3 sub-modules implementation of yolov4 and unit test.

ResBlock has been replaced with direct layer implementation instead of Separate class.
Once the sub-module is available from #7051, we will integrate resBlock into Downsample2 and Downsample3 sub-modules.

### CODE REVIEW

File: models/experimental/functional_yolov4/reference/downsample2.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+# SPDX-License-Identifier: Apache-2.0
+
+
+import torch
+import torch.nn as nn
+from models.experimental.functional_yolov4.reference.resblock import ResBlock
+
+
+class DownSample2(nn.Module):
```
Comment (dvartaniansTT) [inline models/experimental/functional_yolov4/reference/downsample2.py:11]: please generate the pytorch graph of DownSample2 from here and compare to the downSample2 of the reference graph from the original implementation attached here. Also include the auto-generated TTNN graph. and lets ensure all 3 graphs match as a sanity check.

File: models/experimental/functional_yolov4/reference/downsample3.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+# SPDX-License-Identifier: Apache-2.0
+
+
+import torch
+import torch.nn as nn
+from models.experimental.functional_yolov4.reference.resblock import ResBlock
+
+
+class DownSample3(nn.Module):
```
Comment (dvartaniansTT) [inline models/experimental/functional_yolov4/reference/downsample3.py:11]: please generate the pytorch graph of DownSample3 from here and compare to the downSample3 of the reference graph from the original implementation attached here. Also include the auto-generated TTNN graph. and lets ensure all 3 graphs match as a sanity check.

File: models/experimental/functional_yolov4/reference/resblock.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
Comment (dvartaniansTT) [inline models/experimental/functional_yolov4/reference/resblock.py:1]: please generate the pytorch graph of ResBlock from here and compare to the ResBlock of the reference graph from the original implementation attached here. Also include the auto-generated TTNN graph. and lets ensure all 3 graphs match as a sanity check.

File: models/experimental/functional_yolov4/tt/ttnn_downsample2.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import torch.nn as nn
+
+from ttnn.model_preprocessing import preprocess_model
+
+import ttnn
+import tt_lib
+from models.experimental.functional_yolov4.tt.ttnn_resblock import TtResBlock
+
+
+class TtDownSample2:
```
Comment (dvartaniansTT) [inline models/experimental/functional_yolov4/tt/ttnn_downsample2.py:16]: please attached the ttnn graph generated my running the module to this PR so we can ensure it matches with the torch graphs.

File: models/experimental/functional_yolov4/tt/ttnn_downsample3.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import torch.nn as nn
+from models.experimental.functional_yolov4.tt.ttnn_resblock import TtResBlock
+
+from ttnn.model_preprocessing import preprocess_model
+
+import ttnn
+import tt_lib
+
+
+class TtDownSample3:
```
Comment (dvartaniansTT) [inline models/experimental/functional_yolov4/tt/ttnn_downsample3.py:16]: please attached the ttnn graph generated my running the module to this PR so we can ensure it matches with the torch graphs.

File: models/experimental/functional_yolov4/reference/downsample2.py (hunk 1 above)
Comment (dvartaniansTT) [reply models/experimental/functional_yolov4/reference/downsample2.py:11]: [REPO_320x320_YOLOv4_PublicRepo.gv.pdf](https://github.com/tenstorrent-metal/tt-metal/files/14938812/REPO_320x320_YOLOv4_PublicRepo.gv.pdf)

**Review by dvartaniansTT — CHANGES_REQUESTED** [review]

File: models/experimental/functional_yolov4/reference/downsample2.py (hunk 1 above)
Comment (dvartaniansTT) [reply models/experimental/functional_yolov4/reference/downsample2.py:11]: here is a code example of generating the torch graph:

`## Example usage
model = DownSample2()

input_tensor = torch.randn(1, 3, 320, 320)  # Batch size of 1, 3 channels (RGB), 320x320 input - replace this line with the proper input size for your module
output_tensor = model(input_tensor)

model_graph = draw_graph(
    model,
    input_size=(1, 3, 320, 320),
    dtypes=[torch.float32],
    expand_nested=True,
    graph_name="DownSample2",
    depth=100,
    directory=".",
 )
 model_graph.visual_graph.render(format="pdf")`

Comment (keerthana-r-mcw) [reply models/experimental/functional_yolov4/reference/downsample2.py:11]: downsample2_ref
downsample2_ttnn

File: models/experimental/functional_yolov4/reference/downsample3.py (hunk 2 above)
Comment (keerthana-r-mcw) [reply models/experimental/functional_yolov4/reference/downsample3.py:11]: downsample3_ref

File: models/experimental/functional_yolov4/reference/downsample2.py (hunk 1 above)
Comment (dvartaniansTT) [reply models/experimental/functional_yolov4/reference/downsample2.py:11]: > downsample2_ref downsample2_ttnn

I still dont see the pytorch graph for this one! Please use the code snipped I had generated earlier and generate the pytorch graph of your implementation, 
Please **confirm** here that you have generated the torch gerph (attache it here) AND the TTNN graph and compared them with the original graph's DS2 and DS3 portions (I had already attached the original full graph to this PR). 
The purpose of this is to ensure that we have correctly implemented the modules in torch and also to ensure the TTNN graph is a match with torch. 
We can not merge unless these are verified.

**Review by dvartaniansTT — CHANGES_REQUESTED** [review]
Unless we have the 3 graphs (original reference graph, graph of our torch implementation and TTNN graph and we confirm all 3 match (except for mish vs relu), we can not merge this into main.

File: tests/ttnn/integration_tests/yolov4/test_ttnn_downsample2.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+            parameters["c5"], c5_parallel_config = preprocess_conv2d(
+                conv5_weight, conv5_bias, ttnn_module_args.c5, return_parallel_config=True
+            )
+            return parameters
+
+    return custom_preprocessor
+
+
+@skip_for_wormhole_b0()
+def test_downsample2(reset_seeds):
```
Comment (dvartaniansTT) [inline tests/ttnn/integration_tests/yolov4/test_ttnn_downsample2.py:114]: please make this changes:
`def test_downsample2(device, reset_seeds):`
remove the line: `device = ttnn.open_device(device_id=1)`

File: tests/ttnn/integration_tests/yolov4/test_ttnn_downsample2.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    output_tensor = ttnn_model(device, input_tensor)
+
+    #
+    # Tensor Postprocessing
+    #
+    output_tensor = ttnn.to_torch(output_tensor)
+    output_tensor = output_tensor.reshape(1, 80, 80, 128)
+    output_tensor = torch.permute(output_tensor, (0, 3, 1, 2))
+    output_tensor = output_tensor.to(torch_input_tensor.dtype)
+    ttnn.close_device(device)
```
Comment (dvartaniansTT) [inline tests/ttnn/integration_tests/yolov4/test_ttnn_downsample2.py:163]: please remove this line: `ttnn.close_device(device)`

File: tests/ttnn/integration_tests/yolov4/test_ttnn_downsample3.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+                conv5_weight, conv5_bias, ttnn_module_args.c5, return_parallel_config=True
+            )
+
+        return parameters
+
+    return custom_preprocessor
+
+
+@skip_for_wormhole_b0()
+def test_downsample3(reset_seeds):
```
Comment (dvartaniansTT) [inline tests/ttnn/integration_tests/yolov4/test_ttnn_downsample3.py:107]: please apply the same changes I requested for DS2 here as well.

File: tests/ttnn/integration_tests/yolov4/test_ttnn_downsample3.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    output_tensor = ttnn_model(device, input_tensor)
+
+    #
+    # Tensor Postprocessing
+    #
+    output_tensor = ttnn.to_torch(output_tensor)
+    output_tensor = output_tensor.reshape(1, 40, 40, 256)
+    output_tensor = torch.permute(output_tensor, (0, 3, 1, 2))
+    output_tensor = output_tensor.to(torch_input_tensor.dtype)
+    ttnn.close_device(device)
```
Comment (dvartaniansTT) [inline tests/ttnn/integration_tests/yolov4/test_ttnn_downsample3.py:156]: remove this line: `ttnn.close_device(device)`

**Review by dvartaniansTT — CHANGES_REQUESTED** [review]
please fix the issues I have commented.

**Comment by dvartaniansTT** [comment]
please also follow these for testing:

cd $TT_METAL_HOME
scripts/build_scripts/build_with_profiler_opt.sh
./tt_metal/tools/profiler/profile_this.py -n DS1 -c "pytest tests/ttnn/integration_tests/yolov4/test_ttnn_downsample1.py "

File: models/experimental/functional_yolov4/reference/downsample2.py (hunk 1 above)
Comment (keerthana-r-mcw) [reply models/experimental/functional_yolov4/reference/downsample2.py:11]: [downsample2_pytoch.gv.pdf](https://github.com/tenstorrent/tt-metal/files/15006493/downsample2_pytoch.gv.pdf)

File: models/experimental/functional_yolov4/tt/ttnn_downsample3.py (hunk 5 above)
Comment (keerthana-r-mcw) [reply models/experimental/functional_yolov4/tt/ttnn_downsample3.py:16]: [downsample3_pytoch.gv.pdf](https://github.com/tenstorrent/tt-metal/files/15006509/downsample3_pytoch.gv.pdf)

File: models/experimental/functional_yolov4/reference/resblock.py (hunk 3 above)
Comment (keerthana-r-mcw) [reply models/experimental/functional_yolov4/reference/resblock.py:1]: [resblock_downsample2_pytoch.gv.pdf](https://github.com/tenstorrent/tt-metal/files/15006681/resblock_downsample2_pytoch.gv.pdf)
[resblock_downsample3_pytoch.gv.pdf](https://github.com/tenstorrent/tt-metal/files/15006682/resblock_downsample3_pytoch.gv.pdf)

**Comment by keerthana-r-mcw** [comment]
Profiler of downsample3:
[ops_perf_results_DownSample3_2024_04_17_09_43_38.csv](https://github.com/tenstorrent/tt-metal/files/15009189/ops_perf_results_DownSample3_2024_04_17_09_43_38.csv)

Profiler of downsample2:
[ops_perf_results_DownSample2_2024_04_17_09_41_23.csv](https://github.com/tenstorrent/tt-metal/files/15009187/ops_perf_results_DownSample2_2024_04_17_09_41_23.csv)

**Review by dvartaniansTT — APPROVED** [review]

**Review by arakhmati — APPROVED** [review]

**Comment by saichandax** [comment]
Nightly fast dispatch: [passed](https://github.com/tenstorrent/tt-metal/actions/runs/8822843612)
All post commit: [passed](https://github.com/tenstorrent/tt-metal/actions/runs/8822839747)

**Comment by saichandax** [comment]
nightly fast dispatch: [failed](https://github.com/tenstorrent/tt-metal/actions/runs/9008029580)
post commit: [passed](https://github.com/tenstorrent/tt-metal/actions/runs/9008018474)

**Pull request closed without merging**